### DISCUSSION TITLE
question/implement-general-memoization-function

### DISCUSSION THREAD

**Discussion by giscus** [us-fe/us-fe.github.io#13] [Announcements]
# question/implement-general-memoization-function

北美前端面试攻略

https://us-fe.github.io/question/implement-general-memoization-function.html

**Comment by xiaoyu2er** [comment]
https://bigfrontend.dev/problem/implement-general-memoization-function/discuss/5314

**Comment by xiaoyu2er** [comment]
https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Map/Map#%E7%A4%BA%E4%BE%8B

## 语法
new Map([iterable])

## 参数

Iterable 可以是一个数组或者其他 iterable 对象，其元素为键值对(两个元素的数组，例如: [[ 1, 'one' ],[ 2, 'two' ]])。 每个键值对都会添加到新的 Map。null 会被当做 undefined。
## 示例
```js
let myMap = new Map([
  [1, 'one'],
  [2, 'two'],
  [3, 'three'],
])
```

**Comment by mengligit** [comment]
我也是相似的做法，稍有不同的是我的contextMap的外层map的key是存context的

function memo(func, resolver = (...args) => Array.from(args).join('_')) {
   let map = new Map();
   return function (...args) {
      let key = resolver(...args);
      if (map.has(this) && map.get(this).has(key)) {
        return map.get(this).get(key);
      }

      let res = func.apply(this, args);
      if (!map.has(this)) {
        map.set(this, new Map([[key, res]]));
      } else {
        map.get(this).set(key, res);
      }

      return res;
   }
}

**Reply by xiaoyu2er** [reply]
嗯嗯 和我想的一样

只不过假设从第一个思路顺延下来，可能会把 key 当成 cacheKey

**Comment by xiaoyu2er** [comment]
memo这个题思考下如果arguments是1或者”1”的情况下是否也cache了  -- 幽灵蓝谷

**Reply by xiaoyu2er** [reply]
感觉这种特殊的情况 应该是调用者去实现 resolver ?
一个 cacheKey 要考虑这么多情况[捂脸]

**Reply by xiaoyu2er** [reply]
否则memo的 resolver 还要延伸到到 一个复杂类型的key怎么办 一个实例对象的key 怎么办

**Reply by xiaoyu2er** [reply]
你可以说是调用者去实现这个resolver  但是面试时候你就是这个调用者[破涕为笑] -- 幽灵蓝谷

**Reply by xiaoyu2er** [reply]
不同参数类型你都是要去讨论的 -- 幽灵蓝谷

**Reply by xiaoyu2er** [reply]
不知道能不能把 arguments 本身当 key [捂脸]
这要研究下 map key 和 arguments 了

**Reply by xiaoyu2er** [reply]
arguments本身当成key每次都会是一个new reference

**Comment by xiaoyu2er** [comment]
我第一次写memo这道题的时候，直接call了func, 忘记这样写 func.apply(this, arguments); 导致一个test case没过。后来意识到最好call 每个传进来的func都用func.apply(this, arguments)这种方式避免出错。那为啥call resolver的时候不用apply这种写法而是直接 resolver(...arguments)。 我知道这样可以过所有test case,只是想问一个普遍的问题，啥时候需要写成func.apply(this, arguments);

--- 萌

**Reply by xiaoyu2er** [reply]
bfe的test case不是很全面，过了test case不能代表没bug -- stonefish🐠

**Reply by xiaoyu2er** [reply]
memo这题很多人第一个版本写出来是不存this的，那你就可以不用apply，然后面试官就会follow up说如果需要用到怎么办？那你就会说我们得要存this context，存了以后，取出来要用的话，就得要改成apply  - stonefish🐠

**Reply by xiaoyu2er** [reply]
我知道完全和this无关的就不用apply. 问题是怎么快速判断是否与this有关。我第一次写这道题的时候，连func都认为与this无关，根本没想到会有那个test case..... -- 萌

**Reply by xiaoyu2er** [reply]
后来我决定不浪费时间琢磨到底与this有没有关，干脆面试的时候只要是传进来的func一律写成这样func.apply(this, arguments);而不是直接调用 -- 萌

**Reply by xiaoyu2er** [reply]
但有个小细节是不能随便用 arrow
要考虑你所一股脑用的this 是当前函数还是上一层函数
有可能arrow合适有可能不合适

**Reply by xiaoyu2er** [reply]
是的，memo这道题内层函数需要写成传统的function而不是箭头函数，否则this就不对了 -- 萌

**Comment by xiaoyu2er** [comment]
第一个解法

```js
/**
 * @param {Function} func
 * @param {(args:[]) => string }  [resolver] - cache key generator
 */
function memo(func, resolver) {
  // your code here
  const cache = new Map();

  // If my cache key is in my cache, use that value
  // Why use function() over () => {} ?
  // Because function() uses the context from the caller.
  // Arrow functions use lex scoping, so it'll use the context from the memo function.
  return function () {
    const cacheKey = resolver
      ? resolver(...arguments)
      : Array.from(arguments).join(",");

    if (cache.has(cacheKey)) {
      console.log("cached");
      return cache.get(cacheKey);
    }
    // Otherwise invoke the function and add it to my cache
    const val = func.apply(this, arguments);
    cache.set(cacheKey, val);
    return val;
  };
}
```

**Reply by xiaoyu2er** [reply]
存在一个问题就是 this 指向的问题


```js
function fn(a) {
  return `${this.val}_${a}`;
}

const memoFn = memo(fn);

const test1 = {
  val: 1,
  memoFn,
};

const test2 = {
  val: 2,
  memoFn,
};

// 返回 1_1
console.log(test1.memoFn(1));
// 期待返回 2_1 结果被 cache 返回了 1_1
console.log(test2.memoFn(1));
```

**Reply by xiaoyu2er** [reply]
更好的第二个解法

```js
// With this context in mind, check test cases below.
function memo(func, resolver) {
  // your code here
  const cache = new Map();

  // Map<cacheKey, Map<context, value>>
  return function () {
    const cacheKey = resolver
      ? resolver(...arguments)
      : Array.from(arguments).join(",");

    const contextMap = cache.get(cacheKey);
    // If there is a corresponding context map to cachekey
    // Check if context is in the map, if so, return value.
    // Else if no corresponding add contextMap, add new entry to the context map
    if (!contextMap) {
      const value = func.apply(this, arguments);
      cache.set(cacheKey, new Map([[this, value]]));
      return value;
    }

    if (contextMap.has(this)) {
      return contextMap.get(this);
    }
    // If context not in the map, calculate and add to context map.
    const value = func.apply(this, arguments);
    contextMap.set(this, value);
    return value;
  };
}

function testThis(a) {
  return `${this.val}_${a}`;
}

const memoFunc = memo(testThis);

const testSubject = {
  val: 1,
  memo: memoFunc,
};

const testSubject2 = {
  val: 2,
  memo: memoFunc,
};

// 1_1
console.log(testSubject.memo(1));
// Expected no caching and output is 2_1
console.log(testSubject2.memo(1));
// Expected to cache
console.log(testSubject2.memo(1));
```
